[JSC] Remove / AttributeChange should follow to what Structure Add chaining suppression is doing

Constellation opened the pull request:
#### 67c6b7bfcb72ef186c8a9468fecc16b29e1561b1
<pre>
[JSC] Remove / AttributeChange should follow to what Structure Add chaining suppression is doing
<a href="https://bugs.webkit.org/show_bug.cgi?id=283094">https://bugs.webkit.org/show_bug.cgi?id=283094</a>
<a href="https://rdar.apple.com/138491533">rdar://138491533</a>

Reviewed by Yijia Huang.

There is possibility that structure&apos;s properties are already super large (due to
initial capacity), and copying the table causes troubles for memory
consumption and time. This patch adds some guard similar to what we are
doing for Structure add transition to avoid the issue. We pick 4096
based on the currently running benchmarks.

* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::removeNewPropertyTransition):
(JSC::Structure::attributeChangeTransition):
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::shouldDoCacheableDictionaryTransitionForRemoveAndAttributeChange):
(JSC::Structure::transitionCountHasOverflowed const): Deleted.

Canonical link: <a href="https://commits.webkit.org/286601@main">https://commits.webkit.org/286601@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b05cd50746603b203af6e24726ff9f6844e42b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76478 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29384 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81004 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27754 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64655 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3806 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/18083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79545 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49880 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65677 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/40302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47275 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23166 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26077 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69661 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68401 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23498 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82451 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75755 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3854 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2538 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4007 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65646 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/67497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11464 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98009 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11833 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3801 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21442 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3824 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5582 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
r=me

kmiller68 commented on Source/JavaScriptCore/runtime/Structure.h:
```diff
@@ -284,6 +285,22 @@ class Structure : public JSCell {
 
+    inline bool shouldDoCacheableDictionaryTransitionForRemoveAndAttributeChange()
+    {
+        return transitionCountEstimate() > s_maxTransitionLengthForRemove || transitionCountHasOverflowed();
```
Why do we do `transitionCountHasOverflowed()` here but not in `shouldDoCacheableDictionaryTransitionForAdd`?

Constellation replied:
Currently, we are just making the code conservative with this check. We could consider about `transitionCountHasOverflowed` for Add when it is necessary, but so far, we are not observing that so making it conservative choice as this patch's aim is rather avoiding some hugely bad edge cases.

Constellation committed:
[JSC] Remove / AttributeChange should follow to what Structure Add chaining suppression is doing
https://bugs.webkit.org/show_bug.cgi?id=283094
rdar://138491533

Reviewed by Yijia Huang.

There is possibility that structure's properties are already super large (due to
initial capacity), and copying the table causes troubles for memory
consumption and time. This patch adds some guard similar to what we are
doing for Structure add transition to avoid the issue. We pick 4096
based on the currently running benchmarks.

* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::removeNewPropertyTransition):
(JSC::Structure::attributeChangeTransition):
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::shouldDoCacheableDictionaryTransitionForRemoveAndAttributeChange):
(JSC::Structure::transitionCountHasOverflowed const): Deleted.

Canonical link: https://commits.webkit.org/286601@main
